state/event_table.go

```diff
@@ -302,6 +302,54 @@ func (t *EventTable) UpdateBeforeSnapshotID(txn *sqlx.Tx, eventNID, snapID, repl
 	return err
 }
 
+type beforeSnapshotUpdate struct {
+	NID                   int64 `db:"event_nid"`
+	BeforeStateSnapshotID int64 `db:"before_state_snapshot_id"`
+	ReplacesNID           int64 `db:"event_replaces_nid"`
+}
+
+// UpdateBeforeSnapshotIDs sets the before_state_snapshot_id and event_replaces_nid fields for the given NIDs.
+func (t *EventTable) UpdateBeforeSnapshotIDs(txn *sqlx.Tx, updates []beforeSnapshotUpdate) error {
+	chunks := sqlutil.Chunkify2[beforeSnapshotUpdate](3, MaxPostgresParameters, updates)
+
+	for _, chunk := range chunks {
+		// There's quite a lot of clunkiness in the query below.
+		//
+		// 1) We seem to have to explicitly cast types here. According to
+		//        https://github.com/jmoiron/sqlx/issues/796#issuecomment-1026102022,
+		//    "postgres doesn't automatically coerce type" in VALUE expressions.
+		//    We use an explicit CAST expression because the alternative would be a
+		//    quad-colon (https://github.com/jmoiron/sqlx/issues/91#issuecomment-498896062)
+		//    which is a) IMO uglier, and more importantly b) not understood by my IDE.
+		//
+		// 2) The "--)" is a workaround from
+		//        https://github.com/jmoiron/sqlx/issues/828#issuecomment-1214682276
+		//    Without it, passing in 4 updates causes an error
+		//         pq: got 12 parameters but the statement requires 3
+		//    which suggests something silly is happening within sqlx.
+		rows, err := txn.NamedQuery(`
+		UPDATE syncv3_events AS e
+		SET before_state_snapshot_id = u.before_state_snapshot_id,
+		    event_replaces_nid = u.event_replaces_nid
+		FROM (
+			--)
+			VALUES (
+				CAST(:before_state_snapshot_id AS bigint),
+			    CAST(:event_replaces_nid AS bigint),
+				CAST(:event_nid AS bigint)
+			)
+		) AS u(before_state_snapshot_id, event_replaces_nid, event_nid)
+		WHERE e.event_nid = u.event_nid`, chunk)
```

Comment:
I am failing to see what this SQL is doing.

We have a chunk of `beforeSnapshotUpdate` structs. We are somehow making that addressable as `u` by passing in `chunk`, but I'm failing to see any `$1`.

Once we have this addressable as `u`, we just do a basic update which I do understand.

Reply:
Let's drop the `CAST`s for a moment, so that the query is

```sql
		UPDATE syncv3_events AS e
		SET before_state_snapshot_id = u.before_state_snapshot_id,
		    event_replaces_nid = u.event_replaces_nid
		FROM (
			VALUES (:before_state_snapshot_id, :event_replaces_nid, :event_nid)
		) AS u(before_state_snapshot_id, event_replaces_nid, event_nid)
		WHERE e.event_nid = u.event_nid
```

Suppose we have two updates that we want to apply, corresponding to two VALUES expressions `(bss1, replace1, nid1)` and `(bss2, replace2, nid2`). My understanding is that sqlx will transform this query into

```sql
		UPDATE syncv3_events AS e
		SET before_state_snapshot_id = u.before_state_snapshot_id,
		    event_replaces_nid = u.event_replaces_nid
		FROM (
			VALUES ($1, $2, $3), ($4, $5, $6)
		) AS u(before_state_snapshot_id, event_replaces_nid, event_nid)
		WHERE e.event_nid = u.event_nid
```

and pass the query 6 parameters `bss1, replace1, nid1, bss2, replace2, nid2` in that order.

Reply:
https://www.psycopg.org/docs/extras.html#psycopg2.extras.execute_values seems to suggest that it does a similar thing (generate an appropriately-sized VALUES expression). I think psycopg2 does a better job of it, too.